### PR TITLE
Release MFSDP storage after model deletion

### LINKED ISSUE #5616: MFSDP v2 should support HSDP and HFSDP placements
MFSDP v2 currently rejects the placement differences needed for HSDP/HFSDP.

`megatron_fsdp/experimental/parameter_group.py` requires `main_grad` and `main_weight` to have identical placements and raises until HSDP/HFSDP support is implemented. That blocks hybrid sharding modes where weights, gradients, and optimizer state intentionally use different inner-DP / outer-DP layouts.

Implementation detail:

- First microbatch: all-gather across DP-outer.
- Last microbatch: reduce-scatter or all-reduce across DP-outer.
- DP-outer communication should probably run on a separate stream so it can overlap with DP-inner communication.

Expected outcome:

- Define MFSDP v2 layout semantics for HSDP and HFSDP.
- Support HSDP outer-DP replication / gradient synchronization.
- Support HFSDP optimizer-state sharding across inner and outer DP ranks.
- Add focused tests for the required placement and collective transitions.

Development PRs:
- #5652

### PR DESCRIPTION
**Pull request by wujingyue**
MFSDP installed strong back-references from tensor-owned or runtime state to the module graph. Deleting a sharded model could therefore retain its persistent CUDA storage indefinitely.

Make the ownership graph acyclic: parameter ownership markers and post-accumulate hooks hold groups/modules weakly; parameter groups hold their owning modules weakly; the shared context holds its root and prefetch orders weakly; and module hooks dispatch from their runtime module argument rather than close over it. Internal APIs do not expose dereferenced owner or root modules, preventing accidental retention. Deleting a model now releases FSDP storage immediately, without explicit garbage collection or a teardown API.

Add an activation-path regression that runs a sharded model, deletes it, and requires device allocation to return below 1 MiB. It reproduces the leak on origin/main without the main-grad materialization changes.

### CODE REVIEW

**Review by shjwudp — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]

File: tests/unit_tests/distributed/mfsdp_v2/test_fully_shard.py [hunk 1]
```diff
@@ -423,6 +435,28 @@ def test_forward_peak_memory_bounds_in_flight_child_all_gathers(distributed_setu
...
+    model = ElementwiseModel(dim=8192).to(dtype=torch.bfloat16, device=device)
+    fully_shard(model, mesh=mesh, placements=_flat_placements())
+
+    x = torch.ones(1, 8192, dtype=torch.bfloat16, device=device)
+    output = model(x)
+    torch.cuda.synchronize(device)
+    del output, x, model
+    torch.cuda.synchronize(device)
+
+    assert torch.cuda.memory_allocated(device) - allocated_before < 1024**2
```
Comment (skyw) [inline tests/unit_tests/distributed/mfsdp_v2/test_fully_shard.py:456]: Q: is pytorch guaranteed to update `memory_allocated` after dereferencing? regardless whether cache allocator's decision to release the memory or not.

Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_fully_shard.py:456]: Yes, unlike https://docs.pytorch.org/docs/2.13/generated/torch.cuda.memory.memory_reserved.html

**Review by jaredcasper — APPROVED** [review]
